pkg/util/ec2/dmi.go

```diff
@@ -0,0 +1,101 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+package ec2
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/DataDog/datadog-agent/pkg/config"
+	"github.com/DataDog/datadog-agent/pkg/util/dmi"
+	"github.com/google/uuid"
+)
+
+func isBoardVendorEc2() bool {
+	if !config.Datadog.GetBool("ec2_use_dmi") {
+		return false
+	}
+
+	vendor := dmi.GetBoardVendor()
+	if vendor == DMIBoardVendor {
+		return true
+	}
+	return false
+}
+
+// getInstanceIDFromDMI fetches the instance id for current host from DMI
+//
+// On AWS Nitro instances dmi information contains the instanceID for the host. We check that the board vendor is
+// EC2 and that the board_asset_tag match an instanceID format before using it
+func getInstanceIDFromDMI() (string, error) {
+	if !config.Datadog.GetBool("ec2_use_dmi") {
+		return "", fmt.Errorf("'ec2_use_dmi' is disabled")
+	}
+
+	checkEC2UUID := func() {
+		if isEC2UUID() {
+			setCloudProviderSource(metadataSourceUUID)
+		}
+	}
+
+	if !isBoardVendorEc2() {
+		checkEC2UUID()
+		return "", fmt.Errorf("board vendor is not AWS")
+	}
+
+	boardAssetTag := dmi.GetBoardAssetTag()
+	if boardAssetTag == "" || !strings.HasPrefix(boardAssetTag, "i-") {
+		checkEC2UUID()
+		return "", fmt.Errorf("invalid board_asset_tag: '%s'", boardAssetTag)
+	}
+	setCloudProviderSource(metadataSourceDMI)
+	return boardAssetTag, nil
+}
+
+// isEC2UUID returns true if the hypervisor or product UUID starts by "ec2". This doesn't tell us on which instances the
+// agent is running but let us know we're on EC2. Source
+// https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/identify_ec2_instances.html.
+//
+// Depending on the instance type either the DMI product UUID or the hypervisor UUID is available. In both case, if they
+// start with "ec2" we return true.
+func isEC2UUID() bool {
+	uuidData := dmi.GetProductUUID()
+	if uuidData == "" {
+		uuidData = dmi.GetHypervisorUUID()
```

Comment:
Can `dmi.GetHypervisorUUID()` contain `ec2` when `dmi.GetProductUUID()` is not empty and doesn't contain `ec2`?

Reply:
No, both contain the same value but are not always available. Product UUID is available on Nitro instance and require root access, so we can only access it from our container where we run as root.

hypervisor UUID is for non-Nitro instances.